Rename Indexer client to Generator client throughout documentation (#253)
shinzonetwork/docs: content/reference/components/indexer-client/index.md

+++
title = "Generator client"
weight = 2
[extra]
mermaid = true
+++

The Generator client is a standalone Go process that runs as a sidecar alongside an Ethereum validator node. It connects to the validator's Geth node, fetches every new block (with transactions, receipts, and logs), structures them into DefraDB documents, signs the batch, and publishes everything over P2P.

> Generators are write-only data producers. They push data out and reject all incoming replication.

Reviewer: ```suggestion
Generator clients are write-only data producers. They push data out and reject all incoming replication.
```

  DB -- "P2P (libp2p)" --> Hosts
{% end %}

> The generator connects to Geth over two channels:

Reviewer: ```suggestion
The Generator client connects to Geth over two channels:
```

1. Receive: a new block header arrives over the WebSocket subscription.
1. Fetch: the full block is pulled over JSON-RPC, including the header, transactions, receipts, logs, and access lists.
1. Structure: raw data is transformed into the six document types.

> 1. Sign: a Merkle root is computed over all document CIDs and signed with the generator's identity key.

Reviewer: ```suggestion
1. Sign: a Merkle root is computed over all document CIDs and signed with the Generator client's identity key.
```

1. Store: documents are written to the local embedded DefraDB.
1. Publish: DefraDB's P2P layer broadcasts to subscribed peers (hosts).

## Document types

> The generator produces six document types per block. The first four come directly from on-chain data. The last two are metadata that the generator itself produces.

Reviewer: ```suggestion
The Generator client produces six document types per block. The first four come directly from on-chain data. The last two are metadata that the Generator client itself produces.
```

### BlockSignature

> Created after all documents for a block are written. Contains a Merkle root computed over all document CIDs for that block, signed with the generator's identity key.

Reviewer: ```suggestion
Created after all documents for a block are written. Contains a Merkle root computed over all document CIDs for that block, signed with the Generator client's identity key.
```

## Document signing

> Signing is opt-in and configured by the generator operator. Hosts disable it with `--no-signing`.

Reviewer: ```suggestion
Signing is opt-in and configured by the Generator operator. Hosts can disable it with `--no-signing`.
```

The signing flow:

> 1. At startup, the generator loads a persistent identity from the DefraDB keyring using `GetIdentityContext()`.

Reviewer: ```suggestion
1. At startup, the Generator client loads a persistent identity from the DefraDB keyring using `GetIdentityContext()`.
```

1. The identity is injected into the Go context via `node.ContextWithBlockSigning(ctx, collector)`, which enables CID collection.
1. As the block handler writes documents, DefraDB collects the CID of each document written.

> 1. After all documents for a block are written, the generator calls `node.SignBlock`, which computes a Merkle root over the collected CIDs, signs it, and writes the `BlockSignature` document.

Reviewer: ```suggestion
1. After all documents for a block are written, the Generator client calls `node.SignBlock`, which computes a Merkle root over the collected CIDs, signs it, and writes the `BlockSignature` document.
```

Two queue implementations exist:

- GeneratorQueue: tracks document IDs at creation time. When it prunes, it removes entire blocks at once (Block + all its Transactions, Logs, AccessListEntries, and BlockSignature).

> - EventQueue: FIFO queue for P2P replication events. Drains documents as they arrive. Used by hosts, not generator.

Reviewer: ```suggestion
- EventQueue: FIFO queue for P2P replication events. Drains documents as they arrive. Used by Host clients, not Generator clients.
```

## P2P data distribution

> The generator does not manage P2P connections directly. DefraDB handles all of that through libp2p:

Reviewer: ```suggestion
The Generator client does not manage P2P connections directly. DefraDB handles all of that through libp2p:
```

> 1. The generator writes a document.

Reviewer: ```suggestion
1. The Generator client writes a document.
```

1. DefraDB computes a content digest.
1. DefraDB gossips the digest to connected peers.
1. A peer that wants the document requests its full content.
1. DefraDB sends the full document.

> This is unidirectional. The replication filter in `pkg/generator/replication_filter.go` rejects all inbound documents. Generators only push.

Reviewer: ```suggestion
This is unidirectional. The replication filter in `pkg/generator/replication_filter.go` rejects all inbound documents. Generator clients only push.
```
